File: DP/DistinctSubsequences.java
/*
 * LeetCode Problem 115: Distinct Subsequences
 * Problem Number: 115
 * Difficulty: Hard
 * Link: https://leetcode.com/problems/distinct-subsequences/
 * 
 * Given two strings s and t, return the number of distinct subsequences of s which equal t.
 * 
 * Example 1:
 * Input: s = "rabbbit", t = "rabbit"
 * Output: 3
 * Explanation: There are 3 ways to get "rabbit" from "rabbbit":
 *   rabb b it
 *   rab b bit
 *   rab b bit
 * 
 * Example 2:
 * Input: s = "babgbag", t = "bag"
 * Output: 5
 * 
 * Constraints:
 * - 0 <= s.length, t.length <= 1000
 * - s and t consist of English letters.
 * 
 * Topics: String, Dynamic Programming
 * Time Complexity: O(m * n) - where m and n are lengths of s and t
 * Space Complexity: O(m * n) - for the DP table (can be optimized to O(n))
 */

class Solution {
    public int numDistinct(String s, String t) {
        int m = s.length();
        int n = t.length();
        
        // dp[i][j] = number of distinct subsequences of s[0..i-1] that equal t[0..j-1]
        int[][] dp = new int[m + 1][n + 1];
        
        // Empty string t is a subsequence of any s (1 way)
        for (int i = 0; i <= m; i++) {
            dp[i][0] = 1;
        }
        
        // Fill the DP table
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                // Case 1: Exclude current character of s
                dp[i][j] = dp[i - 1][j];
                
                // Case 2: If characters match, include current character of s
                if (s.charAt(i - 1) == t.charAt(j - 1)) {
                    dp[i][j] += dp[i - 1][j - 1];
                }
            }
        }
        
        return dp[m][n];
    }
}

File: Arrays/DistinctSubsequences.go
/*
 * LeetCode Problem 115: Distinct Subsequences
 * Problem Number: 115
 * Difficulty: Hard
 * Link: https://leetcode.com/problems/distinct-subsequences/
 *
 * Given two strings s and t, return the number of distinct subsequences of s which
 * equals t.
 *
 * The test cases are generated so that the answer fits on a 32-bit signed integer.
 *
 *
 *
 * Example 1:
 *
 * Input: s = "rabbbit", t = "rabbit"
 * Output: 3
 * Explanation:
 * As shown below, there are 3 ways you can generate "rabbit" from s.
 * rabbbit
 * rabbbit
 * rabbbit
 *
 * Example 2:
 *
 * Input: s = "babgbag", t = "bag"
 * Output: 5
 * Explanation:
 * As shown below, there are 5 ways you can generate "bag" from s.
 * babgbag
 * babgbag
 * babgbag
 * babgbag
 * babgbag
 *
 *
 *
 * Constraints:
 *
 * 	1 <= s.length, t.length <= 1000
 * 	s and t consist of English letters.
 *
 * Example 1:
 * Input: s = "rabbbit", t = "rabbit"
 * Output: 3
 *
 * Example 2:
 * Input: s = "babgbag", t = "bag"
 * Output: 5
 *
 * Constraints:
 * - 1 <= s.length, t.length <= 1000
 * - s and t consist of English letters.
 *
 * Topics: String, Dynamic Programming
 * Time Complexity: O(n²) typical DP
 * Space Complexity: O(n) or O(n²)
 * Runtime: 0 ms
 * Memory: 42.8 MB
 */

/*
 * Original Solution (java) — translate to Go:
 *
 * class Solution {
 *     public int numDistinct(String s, String t) {
 *         int m=s.length(), n = t.length();
 *         int[][] f= new int[m + 1][n + 1];
 *         for (int i=0;i<m+1; ++i) {
 *             f[i][0]= 1;
 *         }
 *         for (int i= 1; i < m + 1; ++i) {
 *             for (int j= 1; j < n + 1; ++j) {
 *                 f[i][j] = f[i - 1][j];
 *                 if (s.charAt(i - 1) == t.charAt(j - 1)) {
 *                     f[i][j] += f[i-1][j - 1];
 *                 }
 *             }
 *         }
 *         return f[m][n];
 *     }
 * }
 */

package distinctsubsequences

func solve() interface{} {
	// TODO: Implement Go solution
	return nil
}
